File: examples/objective-c/hello_world/ViewController.m
#import <Envoy/Envoy-Swift.h>
#import <UIKit/UIKit.h>
#import "Result.h"
#import "ViewController.h"

#pragma mark - Constants

NSString *_CELL_ID = @"cell-id";
NSString *_REQUEST_AUTHORITY = @"api.lyft.com";
NSString *_REQUEST_PATH = @"/ping";
NSString *_REQUEST_SCHEME = @"https";

#pragma mark - ViewController

@interface ViewController ()
@property (nonatomic, strong) id<StreamClient> client;
@property (nonatomic, strong) NSMutableArray<Result *> *results;
@property (nonatomic, weak) NSTimer *requestTimer;
@end

@implementation ViewController

#pragma mark - Lifecycle

- (instancetype)init {
  self = [super init];
  if (self) {
    self.results = [NSMutableArray new];
    self.tableView.allowsSelection = NO;
    [self startEnvoy];
  }
  return self;
}

- (void)startEnvoy {
  NSLog(@"starting Envoy...");
  NSError *error;
  EngineBuilder *builder = [[EngineBuilder alloc] init];
  id<Engine> engine = [builder buildAndReturnError:&error];
  if (error) {
    NSLog(@"starting Envoy failed: %@", error);
  } else {
    NSLog(@"started Envoy, beginning requests...");
    self.client = [engine getStreamClient];
    [self startRequests];
  }
}

- (void)dealloc {
  [self.requestTimer invalidate];
  self.requestTimer = nil;
}

#pragma mark - Requests

- (void)startRequests {
  self.requestTimer = [NSTimer scheduledTimerWithTimeInterval:1.0
                                                       target:self
                                                     selector:@selector(performRequest)
                                                     userInfo:nil
                                                      repeats:YES];
}

- (void)performRequest {
  NSLog(@"starting request to '%@'", _REQUEST_PATH);

  // Note: this request will use an http/1.1 stream for the upstream request.
  // The Swift example uses h2. This is done on purpose to test both paths in end-to-end tests
  // in CI.
  RequestHeadersBuilder *builder = [[RequestHeadersBuilder alloc] initWithMethod:RequestMethodGet
                                                                          scheme:_REQUEST_SCHEME
                                                                       authority:_REQUEST_AUTHORITY
                                                                            path:_REQUEST_PATH];
  [builder addUpstreamHttpProtocol:UpstreamHttpProtocolHttp1];
  RequestHeaders *headers = [builder build];

  __weak ViewController *weakSelf = self;
  StreamPrototype *prototype = [self.client newStreamPrototype];
  [prototype setOnResponseHeadersWithClosure:^(ResponseHeaders *headers, BOOL endStream) {
    int statusCode = [[[headers valueForName:@":status"] firstObject] intValue];
    NSString *message = [NSString stringWithFormat:@"received headers with status %i", statusCode];
    NSLog(@"%@", message);
    [weakSelf addResponseMessage:message
                    serverHeader:[[headers valueForName:@"server"] firstObject]
                           error:nil];
  }];
  [prototype setOnErrorWithClosure:^(EnvoyError *error) {
    // TODO: expose attemptCount. https://github.com/lyft/envoy-mobile/issues/823
    NSString *message =
        [NSString stringWithFormat:@"failed within Envoy library %@", error.message];
    NSLog(@"%@", message);
    [weakSelf addResponseMessage:message serverHeader:nil error:message];
  }];

  Stream *stream = [prototype startWithQueue:dispatch_get_main_queue()];
  [stream sendHeaders:headers endStream:YES];
}

- (void)addResponseMessage:(NSString *)message
              serverHeader:(NSString *)serverHeader
                     error:(NSString *)error {
  Result *result = [Result new];
  result.message = message;
  result.serverHeader = serverHeader;
  result.error = error;

  [self.results insertObject:result atIndex:0];
  [self.tableView reloadData];
}

#pragma mark - UITableView

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  return self.results.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:_CELL_ID];
  if (cell == nil) {
    cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle
                                  reuseIdentifier:_CELL_ID];
  }

  Result *result = self.results[indexPath.row];
  if (result.error == nil) {
    cell.textLabel.text = result.message;
    cell.detailTextLabel.text =
        [NSString stringWithFormat:@"'server' header: %@", result.serverHeader];

    cell.textLabel.textColor = [UIColor blackColor];
    cell.detailTextLabel.textColor = [UIColor blackColor];
    cell.contentView.backgroundColor = [UIColor whiteColor];
  } else {
    cell.textLabel.text = result.error;
    cell.detailTextLabel.text = nil;

    cell.textLabel.textColor = [UIColor whiteColor];
    cell.detailTextLabel.textColor = [UIColor whiteColor];
    cell.contentView.backgroundColor = [UIColor redColor];
  }

  return cell;
}

@end

File: examples/swift/hello_world/ViewController.swift
import Envoy
import UIKit

private let kCellID = "cell-id"
private let kRequestAuthority = "s3.amazonaws.com"
private let kRequestPath = "/api.lyft.com/static/demo/hello_world.txt"

final class ViewController: UITableViewController {
  private var envoy: Envoy?
  private var requestCount = 0
  private var results = [Result<Response, RequestError>]()
  private var timer: Timer?

  override func viewDidLoad() {
    super.viewDidLoad()
    do {
      self.envoy = try EnvoyBuilder()
        .addLogLevel(.trace)
        .build()
    } catch let error {
      NSLog("Starting Envoy failed: \(error)")
    }

    self.startRequests()
  }

  deinit {
    self.timer?.invalidate()
  }

  // MARK: - Requests

  private func startRequests() {
    // Note that the first delay will give Envoy time to start up.
    self.timer = .scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
      self?.performRequest()
    }
  }

  private func performRequest() {
    self.requestCount += 1
    let requestID = self.requestCount

    NSLog("Starting request to '\(kRequestPath)'")
    let request = RequestBuilder(method: .get, scheme: "http", authority: kRequestAuthority,
                                 path: kRequestPath).build()
    var status = -1
    let handler = ResponseHandler()
      .onHeaders { [weak self] headers, statusCode, _ in
        status = statusCode
        NSLog("Response headers (\(requestID)):\nStatus: \(statusCode)\n\(headers)")
        self?.add(result: .success(Response(id: requestID, body: "",
                                            serverHeader: headers["Server"]?.first ?? "")))
      }
      .onData { data, _ in
        NSLog("Response data (\(requestID)): \(data.count) bytes")
      }
      .onError { [weak self] in
        NSLog("Error (\(requestID)) - request failed")
        self?.add(result: .failure(RequestError(id: requestID, message: "failed, status: \(status)")))
    }

    let stream = self.envoy?.startStream(with: request, handler: handler)
    _ = stream // Not sending additional data up the stream
  }

  private func add(result: Result<Response, RequestError>) {
    self.results.insert(result, at: 0)
    self.tableView.reloadData()
  }

  // MARK: - UITableView

  override func numberOfSections(in tableView: UITableView) -> Int {
    return 1
  }

  override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return self.results.count
  }

  override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath)
    -> UITableViewCell
  {
    let cell = tableView.dequeueReusableCell(withIdentifier: kCellID) ??
      UITableViewCell(style: .subtitle, reuseIdentifier: kCellID)

    let result = self.results[indexPath.row]
    switch result {
    case .success(let response):
      cell.textLabel?.text = "[\(response.id)] \(response.body)"
      cell.detailTextLabel?.text = "'Server' header: \(response.serverHeader)"

      cell.textLabel?.textColor = .black
      cell.detailTextLabel?.textColor = .black
      cell.contentView.backgroundColor = .white

    case .failure(let error):
      cell.textLabel?.text = "[\(error.id)]"
      cell.detailTextLabel?.text = error.message

      cell.textLabel?.textColor = .white
      cell.detailTextLabel?.textColor = .white
      cell.contentView.backgroundColor = .red
    }

    return cell
  }
}
